integration_tests/src/main/python/aqe_test.py

```diff
@@ -0,0 +1,145 @@
+from spark_session import with_cpu_session
+
+_adaptive_conf = { "spark.sql.adaptive.enabled": "true" }
+
+# Dynamic switching of join strategies
+# Dynamic coalescing of shuffle partitions
+# Dynamically Handle Skew Joins
+
+_adaptive_coalese_conf = copy_and_update(_adaptive_conf, {
+    "spark.sql.adaptive.coalescePartitions.enabled": "true"
+})
+
+
+def create_skew_df(spark, length):
+    root = spark.range(0, length)
+    mid = length / 2
+    left = root.select(
+        when(col('id') < mid / 2, mid).
+            otherwise('id').alias("key1"),
+        col('id').alias("value1")
+    )
+    right = root.select(
+        when(col('id') < mid, mid).
+            otherwise('id').alias("key2"),
+        col('id').alias("value2")
+    )
+    return left, right
+
+
+@ignore_order(local=True)
+def test_aqe_skew_join():
+    def do_join(spark):
+        left, right = create_skew_df(spark, 500)
+        left.createOrReplaceTempView("skewData1")
+        right.createOrReplaceTempView("skewData2")
+        return spark.sql("SELECT * FROM skewData1 join skewData2 ON key1 = key2")
+
+    assert_gpu_and_cpu_are_equal_collect(do_join, conf=_adaptive_conf)
+
+@ignore_order(local=True)
+@pytest.mark.parametrize("data_gen", integral_gens, ids=idfn)
+def test_aqe_join_parquet(spark_tmp_path, data_gen):
+    data_path = spark_tmp_path + '/PARQUET_DATA'
+    with_cpu_session(
+        lambda spark: unary_op_df(spark, data_gen).orderBy('a').write.parquet(data_path)
+    )
+
+    def do_it(spark):
+        spark.read.parquet(data_path).createOrReplaceTempView('df1')
+        spark.read.parquet(data_path).createOrReplaceTempView('df2')
+        return spark.sql("select count(*) from df1,df2 where df1.a = df2.a")
+
+    assert_gpu_and_cpu_are_equal_collect(do_it, conf=_adaptive_conf)
+
+
+@ignore_order(local=True)
+@pytest.mark.parametrize("data_gen", integral_gens, ids=idfn)
+def test_aqe_join_parquet_batch(spark_tmp_path, data_gen):
+    # force v2 source for parquet to use BatchScanExec
+    conf = copy_and_update(_adaptive_conf, {
+        "spark.sql.sources.useV1SourceList": ""
+    })
+
+    first_data_path = spark_tmp_path + '/PARQUET_DATA/key=0'
+    with_cpu_session(
+            lambda spark : unary_op_df(spark, data_gen).write.parquet(first_data_path))
+    second_data_path = spark_tmp_path + '/PARQUET_DATA/key=1'
+    with_cpu_session(
+            lambda spark : unary_op_df(spark, data_gen).write.parquet(second_data_path))
+    data_path = spark_tmp_path + '/PARQUET_DATA'
+
+    def do_it(spark):
+        spark.read.parquet(data_path).createOrReplaceTempView('df1')
+        spark.read.parquet(data_path).createOrReplaceTempView('df2')
+        return spark.sql("select count(*) from df1,df2 where df1.a = df2.a")
+
+    assert_gpu_and_cpu_are_equal_collect(do_it, conf=conf)
+
+@ignore_order
+@pytest.mark.parametrize('data_gen', integral_gens, ids=idfn)
+@pytest.mark.parametrize('aqe_enabled', ['true', 'false'], ids=idfn)
```

Comment:
why test with aqe off here?

Reply:
I tested this one on and off because of the window changes. I can remove this here (it was useful for development).